File: sdk/python/src/scion/_pagination.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Pagination helpers for the Scion Python SDK.

Provides ``SyncPage`` and ``AsyncPage`` wrappers that support
automatic iteration across paginated API responses.
"""

from __future__ import annotations

from typing import (
    TYPE_CHECKING,
    Any,
    Callable,
    Generic,
    TypeVar,
)

if TYPE_CHECKING:
    from collections.abc import AsyncIterator, Iterator

T = TypeVar("T")


class SyncPage(Generic[T]):
    """A single page of results from a paginated API endpoint.

    Attributes:
        data: The items on this page.
        has_next: Whether there are more pages available.
        next_cursor: The cursor value for fetching the next page.
    """

    def __init__(
        self,
        data: list[T],
        *,
        has_next: bool = False,
        next_cursor: str | None = None,
        fetch_next: Callable[[str], SyncPage[T]] | None = None,
    ) -> None:
        self.data = data
        self.has_next = has_next
        self.next_cursor = next_cursor
        self._fetch_next = fetch_next

    def __iter__(self) -> Iterator[T]:
        """Iterate over items on this page only."""
        return iter(self.data)

    def __len__(self) -> int:
        return len(self.data)

    def get_next_page(self) -> SyncPage[T]:
        """Fetch the next page of results.

        Raises:
            StopIteration: If there are no more pages.
        """
        if not self.has_next or self.next_cursor is None or self._fetch_next is None:
            raise StopIteration("No more pages")
        return self._fetch_next(self.next_cursor)

    def auto_paging_iter(self) -> Iterator[T]:
        """Iterate over all items across all pages automatically.

        Yields each item, fetching subsequent pages transparently.
        """
        page: SyncPage[T] | None = self
        while page is not None:
            yield from page.data
            if page.has_next and page.next_cursor and page._fetch_next:
                page = page._fetch_next(page.next_cursor)
            else:
                page = None


class AsyncPage(Generic[T]):
    """A single page of results from a paginated async API endpoint.

    Attributes:
        data: The items on this page.
        has_next: Whether there are more pages available.
        next_cursor: The cursor value for fetching the next page.
    """

    def __init__(
        self,
        data: list[T],
        *,
        has_next: bool = False,
        next_cursor: str | None = None,
        fetch_next: Callable[..., Any] | None = None,
    ) -> None:
        self.data = data
        self.has_next = has_next
        self.next_cursor = next_cursor
        self._fetch_next = fetch_next

    def __iter__(self) -> Iterator[T]:
        """Iterate over items on this page only (sync)."""
        return iter(self.data)

    def __len__(self) -> int:
        return len(self.data)

    async def get_next_page(self) -> AsyncPage[T]:
        """Fetch the next page of results asynchronously.

        Raises:
            StopAsyncIteration: If there are no more pages.
        """
        if not self.has_next or self.next_cursor is None or self._fetch_next is None:
            raise StopAsyncIteration("No more pages")
        return await self._fetch_next(self.next_cursor)

    async def auto_paging_iter(self) -> AsyncIterator[T]:
        """Iterate over all items across all pages automatically.

        Yields each item, fetching subsequent pages transparently.
        """
        page: AsyncPage[T] | None = self
        while page is not None:
            for item in page.data:
                yield item
            if page.has_next and page.next_cursor and page._fetch_next:
                page = await page._fetch_next(page.next_cursor)
            else:
                page = None

File: sdk/typescript/src/pagination.ts
/**
 * Pagination helpers for the Scion SDK.
 *
 * Provides the {@link Page} class that wraps a single page of results
 * and implements `AsyncIterable` for transparent auto-pagination via
 * `for await...of`.
 *
 * @packageDocumentation
 */

/**
 * Function signature used by {@link Page} to fetch the next page.
 *
 * @typeParam T - The type of items in the page.
 * @param cursor - The opaque cursor for the next page.
 * @returns A promise resolving to the next Page, or `null` if no more pages.
 */
export type FetchNextPage<T> = (cursor: string) => Promise<Page<T> | null>;

/**
 * A single page of paginated API results.
 *
 * Implements `AsyncIterable<T>` so callers can transparently iterate
 * across all pages using `for await...of`:
 *
 * @example
 * ```ts
 * const firstPage = await client.agents.list();
 * for await (const agent of firstPage) {
 *   console.log(agent.name);
 * }
 * ```
 *
 * @typeParam T - The type of items in this page.
 */
export class Page<T> implements AsyncIterable<T> {
  /** The items on this page. */
  readonly data: T[];

  /** Whether there is a subsequent page. */
  readonly hasNext: boolean;

  /** Opaque cursor to fetch the next page. `undefined` on the last page. */
  readonly nextCursor: string | undefined;

  /** Total number of matching items across all pages (if the server provided it). */
  readonly totalCount: number | undefined;

  /** @internal */
  private readonly fetchNext?: FetchNextPage<T>;

  /**
   * @param data - Items on this page.
   * @param nextCursor - Cursor for the next page, or `undefined` if this is the last page.
   * @param fetchNext - Callback to fetch the next page. Required for auto-pagination.
   * @param totalCount - Total number of matching items (optional).
   */
  constructor(
    data: T[],
    nextCursor: string | undefined,
    fetchNext?: FetchNextPage<T>,
    totalCount?: number,
  ) {
    this.data = data;
    this.nextCursor = nextCursor;
    this.hasNext = nextCursor !== undefined && nextCursor !== '';
    this.fetchNext = fetchNext;
    this.totalCount = totalCount;
  }

  /**
   * Fetch the next page of results.
   *
   * @returns The next Page, or `null` if there are no more pages.
   */
  async getNextPage(): Promise<Page<T> | null> {
    if (!this.hasNext || !this.nextCursor || !this.fetchNext) {
      return null;
    }
    return this.fetchNext(this.nextCursor);
  }

  /**
   * Async iterator that yields every item across this page and all
   * subsequent pages, fetching them lazily.
   */
  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    let page: Page<T> | null = this;
    while (page) {
      for (const item of page.data) {
        yield item;
      }
      page = await page.getNextPage();
    }
  }
}
